rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+            label: Some("Gpu skinning compute shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/src/skinning.wgsl").into()),
+        });
+
+        let pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
+            label: Some("Gpu skinning pipeline"),
+            layout: Some(&layout),
+            module: &module,
+            entry_point: "main",
+        });
+
+        GpuSkinner {
+            vertex_buffers_bgl,
+            joint_matrices_bgl,
+            skinning_inputs_bgl,
+            pipeline,
+        }
+    }
+
+    pub fn execute_pass(
+        &self,
+        device: &Device,
+        encoder: &mut wgpu::CommandEncoder,
+        buffers: &PreSkinningBuffers,
+        mesh_buffers: &MeshBuffers,
+        // The number of inputs in the skinning_inputs buffer
+        skeleton_manager: &SkeletonManager,
+    ) {
+        let vertex_buffers_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&mesh_buffers.vertex_position);
+            bgb.append_buffer(&mesh_buffers.vertex_normal);
+            bgb.append_buffer(&mesh_buffers.vertex_tangent);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_index);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_weight);
+            bgb.build(device, Some("GPU skinning vertex buffers"), &self.vertex_buffers_bgl)
+        };
+
+        let joint_matrices_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&buffers.joint_matrices);
+            bgb.build(device, Some("GPU skinning joint matrices"), &self.joint_matrices_bgl)
+        };
+
+        let skinning_inputs_bgb = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&buffers.gpu_skinning_inputs);
+            bgb.build(device, Some("GPU skinning inputs"), &self.skinning_inputs_bgl)
+        };
+
+        for (i, skel) in skeleton_manager.skeletons().enumerate() {
+            let mut cpass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
+                label: Some(&format_sso!("GPU Skinning {}", i)),
+            });
+            cpass.set_pipeline(&self.pipeline);
+            cpass.set_bind_group(0, &vertex_buffers_bg, &[]);
+            cpass.set_bind_group(1, &joint_matrices_bg, &[]);
+            let offset = (i * std::mem::size_of::<GpuSkinningInput>()) as u32;
+            cpass.set_bind_group(2, &skinning_inputs_bgb, &[offset]);
+
+            let num_verts = (skel.ranges.mesh_range[1] - skel.ranges.mesh_range[0]) as u32;
+            let num_workgroups = (num_verts + (Self::WORKGROUP_SIZE - 1)) / Self::WORKGROUP_SIZE; // Divide rounding up
+            cpass.dispatch(num_workgroups, 1, 1);
+        }
+    }
+}
+
+pub fn add_skinning_to_graph<'node>(
+    graph: &mut RenderGraph<'node>,
+    gpu_skinner: &'node GpuSkinner,
+    pre_skin_data: DataHandle<PreSkinningBuffers>,
+    skinned_data: DataHandle<()>,
```

Comment:
We should give this an empty tag type:

```
struct SkinningOutput;
```

which is about equivalent and as we collect more of these pieces of dummy data will prevent putting the wrong data in the wrong hole.